python/mspasspy/db/database.py

```diff
@@ -1518,229 +1724,188 @@ def _check_required(self, collection='site', keys=['lat','lon','elev','starttime
-                    mspass_object.elog.log_error('update_metadata',
-                            "attribute {} is read only and cannot be updated, but the attribute is saved as READONLYERROR_{}".format(k, k),
-                            ErrorSeverity.Informational)
-                    non_fatal_error_cnt += 1
-                    READONLYERROR_k = "READONLYERROR_" + k
-                    copied_metadata.change_key(k, READONLYERROR_k)
-                    k = READONLYERROR_k
-
-                # save metadata to wf with blocks for write modes ("promiscuous", "cautious", and "pedantic')
-                # promiscuous(no schema check at all)
-                if mode == "promiscuous":
-                    insert_dict[k] = copied_metadata[k]
-                    continue
-
-                # cautious/pedantic(both need schema check)
-                if update_metadata_def.is_defined(k):
-                    # cautious mode: try to fix the required attributes whose types are mismatch with the schema
-                    if mode == "cautious":
-                        # try to convert the mismatch metadata attribute
-                        if not isinstance(copied_metadata[k], update_metadata_def.type(k)):
+                if k in changed_key_list:
+                    if save_schema.is_defined(k):
+                        if not isinstance(copied_metadata[k], save_schema.type(k)):
+                            insertion_dict[k]=copied_metadata[k]
+                        else:
+                            if mode == 'pedantic':
+                                message ='pedantic mode error:  key='+k
+                                if not save_schema.is_defined(k):
+                                    message += ' not defined in schema.'
+                                else:
+                                    value=copied_metadata[k]
+                                    message += ' type of stored value='+str(type(value))+' does not match schema expectation='+str(save_schema.type(k)+'\nAttempting to correct type mismatch')
+                                mspass_object.elog.log_error(alg_name,
+                                    message,ErrorSeverity.Complaint)
+                            # Note we land here for both pedantic and cautious but not promiscuous
                             try:
                                 # The following convert the actual value in a dict to a required type.
                                 # This is because the return of type() is the class reference.
-                                insert_dict[k] = update_metadata_def.type(k)(copied_metadata[k])
+                                oldvalue=copied_metadata[k]
+                                insertion_dict[k] = save_schema.type(k)(copied_metadata[k])
+                                newvalue=copied_metadata[k]
+                                message='Had to convert type of data with key='+k+' from '+str(type(oldvalue))+' to '+str(type(newvalue))
```

Comment:
newvalue here is the same as oldvalue because, in Python, the type cast would not change the original variable.

Reply:
I need your help on that one.   I'm going to check these changes in when I am finished going through your comments.  Could you correct this and teach me what you did?

Reply:
To fix it, we just need to assign `insertion_dict[k]` to `newvalue` because `insertion_dict[k]` is now converted. but `copied_metadata[k]` is still the `oldvalue` one.
That is `newvalue=insertion_dict[k]`.
BTW, we should probably rename the `newvalue` to `new_value`, but that's only my coding style:)